### PR TITLE
Added methods to use native transport if possible

### PR DESCRIPTION
**Pull request by arunrao33**

### CODE REVIEW

**Comment by arunrao33** [comment]
Can i please get a review on this PR? I would love to use it with the epoll support if i could.. thanks.

File: pom.xml [hunk 1]
```diff
@@ -92,6 +92,11 @@
       <artifactId>netty-handler</artifactId>
       <version>${netty.version}</version>
     </dependency>
+    <dependency>
+      <groupId>io.netty</groupId>
+      <artifactId>netty-transport-native-epoll</artifactId>
+      <version>${netty.version}</version>
+    </dependency>
```
Comment (chtyim) [inline pom.xml]: I think you need the classifier. Also, better make this dependency optional.

File: src/main/java/co/cask/http/OsUtils.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.http;
+
+/**
+ * Util class to Detect OS type and other OS related parametera.
+ */
+
```
Comment (chtyim) [inline src/main/java/co/cask/http/OsUtils.java]: Remove this empty line.

File: src/main/java/co/cask/http/OsUtils.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+public class OsUtils {
```
Comment (chtyim) [inline src/main/java/co/cask/http/OsUtils.java]: util class must be final and have a private constructor.

File: src/main/java/co/cask/http/OsUtils.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+    private static final String OS_NAME_KEY = "os.name";
+
+    public static boolean isOSLinux() {
+
```
Comment (chtyim) [inline src/main/java/co/cask/http/OsUtils.java]: Remove empty line.

File: src/main/java/co/cask/http/OsUtils.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        String osName = System.getProperty(OS_NAME_KEY);
+
```
Comment (chtyim) [inline src/main/java/co/cask/http/OsUtils.java]: Also remove this empty line.

File: src/main/java/co/cask/http/NettyHttpService.java [hunk 6]
```diff
@@ -653,6 +706,16 @@ public Builder setExceptionHandler(ExceptionHandler exceptionHandler) {
       return this;
     }
 
+    /**
+      This will set the native transport instead of NIO transport which is more efficient. More info
```
Comment (chtyim) [inline src/main/java/co/cask/http/NettyHttpService.java]: Missing a `*` at the beginning.

File: src/main/java/co/cask/http/NettyHttpService.java [hunk 7]
```diff
@@ -653,6 +706,16 @@ public Builder setExceptionHandler(ExceptionHandler exceptionHandler) {
...
+     * <a href="https://github.com/netty/netty/wiki/Native-transports">native protocol</a>
+     *
```
Comment (chtyim) [inline src/main/java/co/cask/http/NettyHttpService.java]: Remove empty line

File: src/main/java/co/cask/http/NettyHttpService.java [hunk 8]
```diff
@@ -364,6 +368,53 @@ protected void initChannel(SocketChannel ch) throws Exception {
...
+   *
+   * @param workerThreadPoolSize
+   * @return
+   */
+
+  private EventLoopGroup workerGroup(int workerThreadPoolSize) {
+
+    if (this.useNativeTransport && OsUtils.isOSLinux()) {
+      return new EpollEventLoopGroup(workerThreadPoolSize,
+              createDaemonThreadFactory(serviceName + "-worker-thread-%d"));
```
Comment (chtyim) [inline src/main/java/co/cask/http/NettyHttpService.java]: Miss alignment. We align it like this:

```java
return new EpollEventLoopGroup(workerThreadPoolSize,
                               createDaemonThreadFactory(serviceName + "-worker-thread-%d"));
```

File: src/main/java/co/cask/http/NettyHttpService.java [hunk 9]
```diff
@@ -364,6 +368,53 @@ protected void initChannel(SocketChannel ch) throws Exception {
...
+    }
+
+    return new NioEventLoopGroup(workerThreadPoolSize,
+            createDaemonThreadFactory(serviceName + "-worker-thread-%d"));
+  }
+
+
+  private Class selectSocketChannel() {
```
Comment (chtyim) [inline src/main/java/co/cask/http/NettyHttpService.java]: Returns `Class<? extends ServerSocketChannel>` instead.

File: src/main/java/co/cask/http/NettyHttpService.java [hunk 10]
```diff
@@ -364,6 +368,53 @@ protected void initChannel(SocketChannel ch) throws Exception {
...
+
+  /**
+   *
+   * Will create native transport of eventloopgroup if set and possible for acceptor thread.
+   *
+   * @param bossThreadPoolSize
+   * @return
+   */
+  private EventLoopGroup bossGroup(int bossThreadPoolSize) {
+
```
Comment (chtyim) [inline src/main/java/co/cask/http/NettyHttpService.java]: Remove empty line

File: src/main/java/co/cask/http/NettyHttpService.java [hunk 11]
```diff
@@ -364,6 +368,53 @@ protected void initChannel(SocketChannel ch) throws Exception {
...
+  /**
+   *
+   * Will create native transport of eventloopgroup if set and possible for workers.
+   *
+   * @param workerThreadPoolSize
+   * @return
+   */
+
+  private EventLoopGroup workerGroup(int workerThreadPoolSize) {
+
```
Comment (chtyim) [inline src/main/java/co/cask/http/NettyHttpService.java]: Remove empty line

File: src/main/java/co/cask/http/NettyHttpService.java [hunk 12]
```diff
@@ -364,6 +368,53 @@ protected void initChannel(SocketChannel ch) throws Exception {
...
+  /**
+   *
+   * Will create native transport of eventloopgroup if set and possible for workers.
+   *
+   * @param workerThreadPoolSize
+   * @return
+   */
+
+  private EventLoopGroup workerGroup(int workerThreadPoolSize) {
```
Comment (chtyim) [inline src/main/java/co/cask/http/NettyHttpService.java]: This is essentially the same as the `bossGroup` method. Better have one method:

```java
private EventLoopGroup createEventLoopGroup(int size, ThreadFactory threadFactory)
```

File: pom.xml (hunk 1 above)
Comment (arunrao33) [reply pom.xml]: I can make the dependency optional. But adding classifier based on ${os.name} will not build if someone is using non linux OS for development. I use OS X, it fails to find the dependency. Please let me know if there is a different way. I will address all other comments you have.

**Comment by arunrao33** [comment]
@chtyim , please let me know if there any concern with the corrected PR. Thanks.

**Pull request closed without merging**